fix: Not here page when create new request

nkdengineer opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/49801
PROPOSAL: https://github.com/Expensify/App/issues/49801#issuecomment-2378861880


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Case 1
1. Go to workspace chat> + > Submit expense> Complete the flow
2. Pay the request
3. Go to transaction thread> Click on header> Cancel payment
4. Go to main page> + > Submit expense> Complete the flow
5. Navigate to newly created request
6. Verify that: Transaction thread should open


Precondition: Use control workspace and approvals is disabled

#### Case 2
1. Submit expense
2. Pay the expense
3. Cancel the expense
4. In the same workspace chat try to create a new expense
5. Verify that: `Submit expense` and `Create expense` button is not displayed


- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
same as above
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
same as above
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/d382f38b-97ef-4b9b-ab50-8074341b886d


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/c43110b7-f8da-41c6-a622-33b7d0ca8434


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/1455389e-0c7e-41ce-ae7b-442014900716


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/dc0f9af7-b270-445d-9efa-8bc0fcc0d998


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/c68a776d-5c93-476d-b8c6-ee43042cf81d


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/4af430b9-218f-430d-ae43-5c1cddad6f30


</details>

nkdengineer committed:
fix: Not here page when create new request

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
fix: logic merge chat report

nkdengineer commented:
It works well for me.

https://github.com/user-attachments/assets/92e61edf-3a43-4214-83c8-c9a9af90661a

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer commented:
@rushatgabhane i still can't reproduce this bug, please add more details and video screenshot

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer commented:
@rushatgabhane friendly bump

nkdengineer commented:
@rushatgabhane Friendly bump.

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer commented:
Can reproduce now.

nkdengineer commented:
@tgolen Another problem here, in `OpenReport` API, `iouReportID` of the policy expense report is returned as `null`.

<img width="1512" alt="Screenshot 2024-11-12 at 15 33 07" src="https://github.com/user-attachments/assets/43d7c457-cb84-4f7b-9df9-bfde7a933ed3">

tgolen commented:
OK, thanks! Can you please report that using the template here: https://github.com/Expensify/App/issues/new?assignees=&labels=Hot+Pick%2C+Daily%2C+Internal%2C+AutoAssignerNewDotQuality&projects=&template=Internal.md and then someone can get that looked into

nkdengineer commented:
@tgolen This problem is blocking us from testing this PR and aslo related to this issue can you handle this?

tgolen commented:
I can, yes. It would still be nice if you could use that template and then I can just assign it to myself. I'm not 100% clear on what the problem is or how to reproduce it. This one is a bit different than the one I previously fixed (`parentReportID`)

nkdengineer commented:
@tgolen Created the new issue [here](https://github.com/Expensify/App/issues/52527).

nkdengineer commented:
@tgolen Is the backend fix done?

tgolen commented:
I _think_ so. I see [this backend issue](https://github.com/Expensify/Expensify/issues/411558) was closed and the PR fixing it has been deployed. Could you see if you're still able to reproduce the problem?

As a reminder, the backend bug that was done that should fix this was a bug in this step here:

![image](https://github.com/user-attachments/assets/57333e38-ba88-4de7-bb60-8c0ab2b60c7e)

nkdengineer commented:
Thanks, let me check again.

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer commented:
@tgolen It's still not fixed, `OpenReport` still return `iouReportID` as `null` for policy expense report.

https://github.com/user-attachments/assets/ba2ca0d2-d59d-483c-8fc2-f8251604d282

tgolen commented:
Just to verify, you went through the entire setup flow with a brand new workspace, and could still reproduce it? The fix that was applied wouldn't work for an existing setup.

nkdengineer commented:
> Just to verify, you went through the entire setup flow with a brand new workspace, and could still reproduce it? The fix that was applied wouldn't work for an existing setup.

@tgolen  Yes, I tested with a brand-new workspace. The problem I want to mention here is

1. `CancelPayment` API returned correctly `iouReportID` of policy expense chat.
2. But after that, `OpenReport` API returns it as `null`.

Is `iouReportID` of policy expense chat isn't updated to DB in `CancelPayment` API

tgolen commented:
@nkdengineer I agree with you that the `iouReportID` is still not being returned from `OpenReport`. However, I spoke with @youssef-lr who said that should be OK.

When I complete the reproduction steps, I'm able to submit more expenses to the workspace after the admin has cancelled a payment. So, I think the initial bug reported [here](https://github.com/Expensify/App/pull/50847#issuecomment-2433453547) by @rushatgabhane is indeed fixed.

Can you try to verify that, please?

youssef-lr commented:
Can you show a recording of this bug @nkdengineer?

tgolen commented:
@nkdengineer Here is a video of me going through the full cancel expense flow the way that I understand it (there are some duplicate expense violations showing, but I think that's OK). Maybe you can spot if I am misunderstanding the reproduction steps and the bug. If so, can you please outline the proper steps?


https://github.com/user-attachments/assets/26ae7058-09dd-4803-8bd7-9082ee331374

youssef-lr commented:
My guess is that in the recording from Rushat, the `iouReportID` didn't get updated to `null` and the cancelled report was still its `iouReportID`, so the app tried to add an expense to it which is not possible (and expected). @nkdengineer are you currently still able to reproduce this?

tgolen commented:
Here is my reproduction without any approval workflow, and the second expense is still successful. I believe this follow's Rush's steps, but those steps are also pretty bare, so I hope I got it right.

https://github.com/user-attachments/assets/084334a5-8e17-4d70-a66e-0e9dfb55a94a

nkdengineer commented:
> @youssef-lr It looks like with approvals turned off, the canceled reports has status=1 state=1. The way I understand it, this should still allow more expenses to be added to it.

@tgolen   Yeah, we're blocking this case because of the bug [here](https://github.com/Expensify/App/pull/50847#issuecomment-2494432981).

youssef-lr commented:
@nkdengineer please post a recording of the bug with approvals disabled and attempt to create a new expense from the workspace chat after cancelling the last report.

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer commented:
@youssef-lr Here is the bug and here is the [RCA](https://github.com/Expensify/App/pull/50847#issuecomment-2494432981)


https://github.com/user-attachments/assets/a1ac5e65-42ae-41ca-9dd9-25d718867541

youssef-lr commented:
@nkdengineer can you try something please,  do "Clear cache and restart" then request again from that same report?

youssef-lr commented:
One more request @nkdengineer can you please the response of `CancelPayment` from the network tab?

nkdengineer commented:
```
{
    "jsonCode": 200,
    "requestID": "8e6b92cb59b88b75-HKG",
    "onyxData": [
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "stateNum": 2
            }
        },
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "statusNum": 2
            }
        },
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "lastReadTime": "2024-11-22 20:12:16.677"
            }
        },
        {
            "key": "report_1634788530390232",
            "onyxMethod": "merge",
            "value": {
                "lastActorAccountID": 18725224,
                "lastMessageText": "Nkdengineer+ididk@outlook.com's Workspace owes ₫1,677,895",
                "lastVisibleActionCreated": "2024-11-22 19:18:42.237",
                "maxSequenceNumber": null,
                "reportID": "1634788530390232"
            }
        },
        {
            "key": "reportActions_1634788530390232",
            "onyxMethod": "merge",
            "shouldNotify": false,
            "value": {
                "5574639573521203775": {
                    "actionName": "REPORTPREVIEW",
                    "actorAccountID": 18725224,
                    "automatic": false,
                    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/default-avatar_17.png",
                    "childCommenterCount": 1,
                    "childLastActorAccountID": 18725224,
                    "childLastVisibleActionCreated": "2024-11-22 19:56:30.617",
                    "childManagerAccountID": 18725224,
                    "childMoneyRequestCount": 2,
                    "childOldestFourAccountIDs": "18725224",
                    "childOwnerAccountID": 18725224,
                    "childReportID": "5560405953103041",
                    "childReportName": "Expense Report #5560405953103041",
                    "childReportNotificationPreference": "hidden",
                    "childStateNum": 2,
                    "childStatusNum": 2,
                    "childType": "expense",
                    "childVisibleActionCount": 1,
                    "created": "2024-11-22 19:18:42.237",
                    "isAttachment": false,
                    "lastModified": "2024-11-22 19:18:42.237",
                    "message": [
                        {
                            "deleted": "",
                            "html": "Nkdengineer+ididk@outlook.com's Workspace owes ₫1,677,895",
                            "isDeletedParentAction": false,
                            "isEdited": false,
                            "reactions": [],
                            "text": "Nkdengineer+ididk@outlook.com's Workspace owes ₫1,677,895",
                            "type": "COMMENT",
                            "whisperedTo": []
                        }
                    ],
                    "originalMessage": {
                        "isNewDot": true,
                        "lastModified": "2024-11-22 19:18:42.237",
                        "linkedReportID": "5560405953103041"
                    },
                    "person": [
                        {
                            "style": "strong",
                            "text": "nkdengineer+ididk@outlook.com",
                            "type": "TEXT"
                        }
                    ],
                    "reportActionID": "5574639573521203775",
                    "reportActionTimestamp": 1732303122237,
                    "sequenceNumber": null,
                    "shouldShow": true,
                    "timestamp": 1732303122,
                    "whisperedToAccountIDs": []
                }
            }
        },
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "avatarUrl": null,
                "chatReportID": "1634788530390232",
                "chatType": "",
                "currency": "VND",
                "description": "",
                "errorFields": {
                    "export": null,
                    "notFound": null
                },
                "fieldList": null,
                "hasOutstandingChildRequest": false,
                "hasOutstandingChildTask": false,
                "hasParentAccess": true,
                "invoiceReceiver": null,
                "iouReportID": null,
                "isCancelledIOU": false,
                "isDeletedParentAction": null,
                "isOwnPolicyExpenseChat": false,
                "isPinned": false,
                "isWaitingOnBankAccount": false,
                "lastActionType": "REIMBURSEMENTDEQUEUED",
                "lastActorAccountID": "18725224",
                "lastMessageHtml": "",
                "lastMessageText": "",
                "lastReadSequenceNumber": 0,
                "lastVisibleActionCreated": "2024-11-22 20:12:16.677",
                "lastVisibleActionLastModified": "2024-11-22 20:12:16.677",
                "managerID": 18725224,
                "nonReimbursableTotal": 0,
                "oldPolicyName": "",
                "ownerAccountID": 18725224,
                "parentReportActionID": "5574639573521203775",
                "parentReportID": "1634788530390232",
                "participants": {
                    "18725224": {
                        "notificationPreference": "hidden"
                    }
                },
                "permissions": [],
                "policyAvatar": null,
                "policyID": "02F5C9ED7AB9446E",
                "policyName": null,
                "private_isArchived": "",
                "reportID": "5560405953103041",
                "reportName": "Expense Report #5560405953103041",
                "stateNum": 2,
                "statusNum": 2,
                "total": -167789500,
                "tripData": null,
                "type": "expense",
                "unheldNonReimbursableTotal": 0,
                "unheldTotal": -167789500,
                "visibility": null,
                "welcomeMessage": "",
                "writeCapability": "all"
            }
        },
        {
            "key": "reportNameValuePairs_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "type": "expense"
            }
        },
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "isCancelledIOU": true
            }
        },
        {
            "key": "report_1634788530390232",
            "onyxMethod": "merge",
            "value": {
                "iouReportID": 5560405953103041
            }
        },
        {
            "key": "report_5560405953103041",
            "onyxMethod": "merge",
            "value": {
                "lastActorAccountID": 18725224,
                "lastMentionedTime": null,
                "lastMessageText": "You",
                "lastReadTime": "2024-11-22 20:12:16.677",
                "lastVisibleActionCreated": "2024-11-22 20:12:16.677",
                "maxSequenceNumber": 12,
                "reportID": "5560405953103041"
            }
        },
        {
            "key": "reportActions_5560405953103041",
            "onyxMethod": "merge",
            "shouldNotify": false,
            "value": {
                "6037595170825844431": {
                    "actionName": "REIMBURSEMENTDEQUEUED",
                    "actorAccountID": 18725224,
                    "automatic": false,
                    "avatar": "https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/default-avatar_17.png",
                    "created": "2024-11-22 20:12:16.677",
                    "isAttachment": false,
                    "lastModified": "2024-11-22 20:12:16.677",
                    "message": [
                        {
                            "style": "strong",
                            "text": "You",
                            "type": "TEXT"
                        },
                        {
                            "style": "normal",
                            "text": " cancelled the reimbursement",
                            "type": "TEXT"
                        }
                    ],
                    "originalMessage": {
                        "amount": 167789500,
                        "cancellationReason": "CANCEL_REASON_ADMIN",
                        "currency": "VND",
                        "expenseReportID": 5560405953103041,
                        "isNewDot": true,
                        "lastModified": "2024-11-22 20:12:16.677"
                    },
                    "person": [
                        {
                            "style": "strong",
                            "text": "nkdengineer+ididk@outlook.com",
                            "type": "TEXT"
                        }
                    ],
                    "reportActionID": "6037595170825844431",
                    "reportActionTimestamp": 1732306336677,
                    "sequenceNumber": 12,
                    "shouldShow": true,
                    "timestamp": 1732306336,
                    "whisperedToAccountIDs": []
                }
            }
        }
    ],
    "previousUpdateID": 3075688582,
    "lastUpdateID": 3075689746
}
```

nkdengineer commented:
@youssef-lr Tried with clear cache and restart.

https://github.com/user-attachments/assets/a216fd09-677e-4dba-9a2d-5476d9d5f8fc

youssef-lr commented:
Thanks! we clearly have an issue in App, we should allow adding expenses to a report that isn't in state/status 1, 1. The report you have there has state/status 2,2. The `iouReportID` of the workspace chat returned from `CancelPayment` also should be `null`.

youssef-lr commented:
@nkdengineer please add this update to this PR to prevent showing the "Create expense" when we can't add expenses anymore. The backend is right here.

<img width="1330" alt="Screenshot 2024-11-22 at 21 29 01" src="https://github.com/user-attachments/assets/e87c1cb2-0f59-454b-a618-29ec7c0a11b4">

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
fix: not show create expense button when it closed

nkdengineer commented:
https://github.com/user-attachments/assets/7ee238de-9ed5-4078-a760-378e9230a3cd

@rushatgabhane @youssef-lr i updated the code and added steps in OP

tgolen commented on src/libs/actions/IOU.ts:
```diff
@@ -7328,6 +7332,13 @@ function cancelPayment(expenseReport: OnyxEntry<OnyxTypes.Report>, chatReport: O
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${chatReport.reportID}`,
+            value: {
+                iouReportID: (iouCreatedAction?.created ?? '') > (expenseCreatedAction?.created ?? '') ? chatReport?.iouReportID : expenseReport.reportID,
```
Can you please add a comment to the code that explains why this `>` logic is here and why `iouReportID` should be used in this case?

tgolen reviewed (changes requested):
Could you also please add a unit test for this?

nkdengineer committed:
add comment

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
fix: add test for cancel payment

nkdengineer commented:
I added the test but it seems to be failing test 1, will check today

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
Add test for cancel payment

nkdengineer commented:
@rushatgabhane @tgolen i added the tests, please check again

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,108 @@ describe('actions/IOU', () => {
+            let expenseReport: OnyxEntry<OnyxTypes.Report>;
+            let chatReport: OnyxEntry<OnyxTypes.Report>;
+
+            mockFetch?.pause?.();
```
Why is this paused?

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,108 @@ describe('actions/IOU', () => {
+                .then(
+                    () =>
+                        new Promise<void>((resolve) => {
+                            const connection = Onyx.connect({
+                                key: ONYXKEYS.COLLECTION.REPORT,
+                                waitForCollectionCallback: true,
+                                callback: (allReports) => {
+                                    Onyx.disconnect(connection);
+                                    expenseReport = Object.values(allReports ?? {}).find((report) => report?.type === CONST.REPORT.TYPE.IOU);
+
+                                    resolve();
+                                },
+                            });
```
This seems to be a pretty repeatable pattern:
- Connect to an onyx key
- Disconnect immediately from the connection
- Get a piece of data

How about adding a utility to https://github.com/Expensify/App/blob/main/tests/utils/TestHelper.ts which would simplify this like:

```
.then(() => getOnyxData(ONYXKEYS.COLLECTION.REPORT, (allReports) => {
    expenseReport = Object.values(allReports ?? {}).find((report) => report?.type === CONST.REPORT.TYPE.IOU);
}))
```
```

nkdengineer replied:
@tgolen Updated with new util `getOnyxData`.

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,108 @@ describe('actions/IOU', () => {
+                    return waitForBatchedUpdates();
+                })
+                .then(() => {
+                    mockFetch?.pause?.();
```
Why is it paused here again?

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
fix: add comment for cancel payment test

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
implement getOnyxData function

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+            let expenseReport: OnyxEntry<OnyxTypes.Report>;
+            let chatReport: OnyxEntry<OnyxTypes.Report>;
+
+            Onyx.set(ONYXKEYS.SESSION, {email: CARLOS_EMAIL, accountID: CARLOS_ACCOUNT_ID});
+            return waitForBatchedUpdates()
+                .then(() => {
+                    PolicyActions.createWorkspace(CARLOS_EMAIL, true, "Carlos's Workspace");
+                    return waitForBatchedUpdates();
+                })
+                .then(
+                    () =>
+                        new Promise<void>((resolve) => {
+                            const connection = Onyx.connect({
+                                key: ONYXKEYS.COLLECTION.REPORT,
+                                waitForCollectionCallback: true,
+                                callback: (allReports) => {
+                                    Onyx.disconnect(connection);
+                                    chatReport = Object.values(allReports ?? {}).find((report) => report?.chatType === CONST.REPORT.CHAT_TYPE.POLICY_EXPENSE_CHAT);
+
+                                    resolve();
+                                },
+                            });
+                        }),
+                )
+                .then(() => {
+                    if (chatReport) {
+                        // When a submit IOU expense is made
```
```suggestion
            // Given a signed in account, which owns a workspace, and has a policy expense chat
            Onyx.set(ONYXKEYS.SESSION, {email: CARLOS_EMAIL, accountID: CARLOS_ACCOUNT_ID});
            return waitForBatchedUpdates()
                .then(() => {
                    // Which owns a workspace
                    PolicyActions.createWorkspace(CARLOS_EMAIL, true, "Carlos's Workspace");
                    return waitForBatchedUpdates();
                })
                .then(
                    () =>
                        // And
                        new Promise<void>((resolve) => {
                            const connection = Onyx.connect({
                                key: ONYXKEYS.COLLECTION.REPORT,
                                waitForCollectionCallback: true,
                                callback: (allReports) => {
                                    Onyx.disconnect(connection);
                                    chatReport = Object.values(allReports ?? {}).find((report) => report?.chatType === CONST.REPORT.CHAT_TYPE.POLICY_EXPENSE_CHAT);

                                    resolve();
                                },
                            });
                        }),
                )
                .then(() => {
                    if (chatReport) {
                        // When an IOU expense is submitted to that policy expense chat
```

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+                    return waitForBatchedUpdates();
+                })
+                .then(() =>
+                    TestHelper.getOnyxData({
```
```suggestion
                    // And given an expense report has now been created which holds the IOU
                    TestHelper.getOnyxData({
```

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+                    }),
+                )
+                .then(() => {
+                    if (chatReport && expenseReport) {
```
```suggestion
                    // When the expense report is paid elsewhere (but really, any payment option would work)
                    if (chatReport && expenseReport) {
```

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+                })
+                .then(() => {
+                    if (chatReport && expenseReport) {
+                        // When we attempt to cancle payment an expense from the expense report
```
```suggestion
                        // And when the payment is cancelled
```

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+                            const connection = Onyx.connect({
+                                key: ONYXKEYS.COLLECTION.REPORT,
+                                waitForCollectionCallback: true,
+                                // When fetching all reports from Onyx
```
```suggestion
```

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,104 @@ describe('actions/IOU', () => {
+                                callback: (allReports) => {
+                                    Onyx.disconnect(connection);
+                                    const chatReportData = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${chatReport?.reportID}`];
+                                    // Then we should find an IOU action with specific properties
```
```suggestion
                                    // Then the policy expense chat report has the iouReportID of the IOU expense report
```

tgolen reviewed (changes requested):
The util looks good! Here are some suggestions to improve the comments.

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer committed:
Update tests/actions/IOUTest.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

nkdengineer commented:
@tgolen Updated.

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,108 @@ describe('actions/IOU', () => {
+                    () =>
+                        // And
+                        new Promise<void>((resolve) => {
+                            const connection = Onyx.connect({
```
Use `getOnyxData()` instead

tgolen commented on tests/actions/IOUTest.ts:
```diff
@@ -1907,6 +1907,108 @@ describe('actions/IOU', () => {
+                .then(
+                    () =>
+                        new Promise<void>((resolve) => {
+                            const connection = Onyx.connect({
```
Here too

tgolen reviewed (changes requested):
Oops, sorry I didn't catch these earlier. The comments look better now!

nkdengineer committed:
Merge branch 'main' into fix/49801

nkdengineer committed:
use getOnyxData

nkdengineer commented:
@tgolen All done.

Do you think we should create a new issue to update all the same instances in the test to use `getOnyxData` helper? If yes, I'd like to work on this.

tgolen reviewed (approved):
In general, I think the cleanup would be nice, but it's very low in value and priority. Just doing a quick search I see about 187 instances of `const connection = Onyx.connect({` in the test files.

I think trying to use the new utility when you can and pointing it out when reviewing PRs is a good place to start for now.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
